src/ArborX_DBSCAN.hpp

```diff
@@ -126,57 +225,202 @@ dbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
+      bvh.query(exec_space, predicates,
+                Details::CountUpToN<MemorySpace>{num_neigh, core_min_size});
+      Kokkos::Profiling::popRegion();
+      elapsed["neigh"] = timer_seconds(timer_local);
+
+      using CorePoints = DBSCAN::DBSCANCorePoints<MemorySpace>;
+
+      // Perform the queries and build clusters through callback
+      timer_start(timer_local);
+      Kokkos::Profiling::pushRegion("ArborX::dbscan::clusters:query");
+      bvh.query(exec_space, predicates,
+                Details::FDBSCANCallback<MemorySpace, CorePoints>{
+                    labels, CorePoints{num_neigh, core_min_size}});
+      Kokkos::Profiling::popRegion();
+      elapsed["query"] = timer_seconds(timer_local);
+    }
   }
-  else
+  else if (parameters._implementation ==
+           DBSCAN::Implementation::FDBSCAN_DenseBox)
   {
-    // Determine core points
-    Kokkos::Timer timer_local;
-    timer_start(timer_local);
-    Kokkos::Profiling::pushRegion("ArborX::dbscan::clusters::num_neigh");
-    Kokkos::View<int *, MemorySpace> num_neigh("ArborX::dbscan::num_neighbors",
-                                               n);
-    bvh.query(exec_space, predicates,
-              Details::CountUpToN<MemorySpace>{num_neigh, core_min_size});
-    Kokkos::Profiling::popRegion();
-    elapsed["neigh"] = timer_seconds(timer_local);
+    // Find dense boxes
+    timer_start(timer);
+    Kokkos::Profiling::pushRegion("ArborX::dbscan::dense_cells");
+    Box bounds;
+    Details::TreeConstruction::calculateBoundingBoxOfTheScene(
+        exec_space, primitives, bounds);
+
+    // The cell length is chosen to be eps/sqrt(dimension), so that any two
+    // points within the same cell are within eps distance of each other.
+    float const h = eps / std::sqrt(3); // 3D (for 2D change to std::sqrt(2))
+    Details::CartesianGrid const grid(bounds, h);
+
+    auto cell_indices =
+        Details::computeCellIndices(exec_space, primitives, grid);
+
+    auto permute = Details::sortObjects(exec_space, cell_indices);
+    auto &sorted_cell_indices = cell_indices; // alias
+
+    int num_nonempty_cells;
+    int num_points_in_dense_cells;
+    {
+      // Reorder indices and permutation so that the dense cells go first
+      auto cell_offsets =
+          Details::computeOffsetsInOrderedView(exec_space, sorted_cell_indices);
+      num_nonempty_cells = cell_offsets.size() - 1;
+
+      num_points_in_dense_cells = Details::reorderDenseAndSparseCells(
+          exec_space, cell_offsets, core_min_size, sorted_cell_indices,
+          permute);
+    }
+    int num_points_in_sparse_cells = n - num_points_in_dense_cells;
 
-    using CorePoints = DBSCAN::DBSCANCorePoints<MemorySpace>;
+    auto dense_sorted_cell_indices = Kokkos::subview(
+        sorted_cell_indices, Kokkos::make_pair(0, num_points_in_dense_cells));
+
+    auto dense_cell_offsets = Details::computeOffsetsInOrderedView(
+        exec_space, dense_sorted_cell_indices);
```

Comment:
Not sure why this is needed.  I think I understand the rest.

Reply:
After i reordered cell indices to put dense cells first, current cell offsets are no longer valid. I will still need offsets in the callbacks, so i need to recompute them.